src/mame/sun/v120.cpp

```diff
@@ -0,0 +1,159 @@
+#include "emu.h"
+
+#include "cpu/h8/h83437.h"
+#include "bus/rs232/rs232.h"
+#include "machine/input_merger.h"
+#include "machine/i2cmem.h"
+
+#include "speaker.h"
+
+namespace {
+
+class v120_state : public driver_device
+{
+public:
+	v120_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag),
+		m_flashv(0x11),
+		m_lom(*this, "lom"),
+		m_ttya(*this, "ttya"),
+		m_lom_eeprom(*this, "lom_eeprom")
+	{ }
+
+	void v120(machine_config &config);
+
+private:
+	u8 m_flashv;
+
+	required_device<h83437_device> m_lom;
+	required_device<rs232_port_device> m_ttya;
+	required_device<i2cmem_device> m_lom_eeprom;
+
+	void mcu_map(address_map &map);
+	auto sda_read();
+
+	void lom_set_i2c(int state);
+};
+
+void v120_state::mcu_map(address_map &map)
+{
+	map(0x0000, 0xf77f).rom().region("lom", 0);
+
+	// Kludge! Needed sp that the newly booted firmware doesn't think it just
+	// upgraded. Probably needs to be removed altogether with a real ROM dump
+	// as opposed to one from firmware upgrade package.
+	map(0xf77e, 0xf77e).lrw8(NAME([this]() { return m_flashv; }), NAME([this](u8 data) { m_flashv = data; }));
+}
+
+static DEVICE_INPUT_DEFAULTS_START(terminal)
+	DEVICE_INPUT_DEFAULTS("RS232_RXBAUD", 0xff, RS232_BAUD_9600)
+	DEVICE_INPUT_DEFAULTS("RS232_TXBAUD", 0xff, RS232_BAUD_9600)
+	DEVICE_INPUT_DEFAULTS("RS232_DATABITS", 0xff, RS232_DATABITS_8)
+	DEVICE_INPUT_DEFAULTS("RS232_PARITY", 0xff, RS232_PARITY_NONE)
+	DEVICE_INPUT_DEFAULTS("RS232_STOPBITS", 0xff, RS232_STOPBITS_1)
+DEVICE_INPUT_DEFAULTS_END
+
+auto v120_state::sda_read()
+{
+	u8 lom_eeprom_sda = m_lom_eeprom->read_sda();
+	logerror ("sda_read: eeprom=%d\n", lom_eeprom_sda);
+
+	// SDA lines from other devices will be mixed in here
+	return lom_eeprom_sda;
+}
+
+void v120_state::v120(machine_config &config)
+{
+	// Lights-Out Management controller.
+	H83437(config, m_lom, XTAL(14'318'181));
+	m_lom->set_addrmap(AS_PROGRAM, &v120_state::mcu_map);
+	m_lom->write_sci_tx<1>().set(m_ttya, FUNC(rs232_port_device::write_txd));
+
+	RS232_PORT(config, m_ttya, default_rs232_devices, "terminal");
+	m_ttya->rxd_handler().set(m_lom, FUNC(h83437_device::sci_rx_w<1>));
+	m_ttya->set_option_device_input_defaults("terminal", DEVICE_INPUT_DEFAULTS_NAME(terminal));
+
+	// Debug chatter
+	m_lom->scl_cb().set([this](int state) { logerror("scl_write %d\n", state); });
```

Comment:
The usual convention is to group related device functions. In this case everything handling `m_lom` might be better put above where the CPU is configured.